### PR TITLE
Fixed #4206 and #4207 open and close notebook quickly issue #4240

### LINKED ISSUE #4206: TypeError: Cannot read property '0' of undefined
- Azure Data Studio Version:

Steps to Reproduce:

1. open and close sql notebook few times
2. got exception:
    TypeError: Cannot read property '0' of undefined
    at NotebookModel.setProviderIdForKernel (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:885:77)
    at NotebookModel.doChangeKernel (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:584:18)
    at NotebookModel.loadKernelInfo (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:687:26)
    at NotebookModel.<anonymous> (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:513:34)
    at step (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:48:23)
    at Object.next (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:29:53)
    at file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:23:71
    at new ZoneAwarePromise (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:891:29)
    at __awaiter (file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:19:12)
    at file:///D:/3/azuredatastudio/out/sql/parts/notebook/models/notebookModel.js:505:92
    at ZoneDelegate.invoke (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:388:26)
    at Object.onInvoke (d:\3\azuredatastudio\node_modules\@angular\core\bundles\core.umd.js:4156:37)
    at ZoneDelegate.invoke (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:387:32)
    at Zone.run (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:138:43)
    at D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:872:34
    at ZoneDelegate.invokeTask (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:421:31)
    at Object.onInvokeTask (d:\3\azuredatastudio\node_modules\@angular\core\bundles\core.umd.js:4147:37)
    at ZoneDelegate.invokeTask (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:420:36)
    at Zone.runTask (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:188:47)
    at drainMicroTaskQueue (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:595:35)
    at ZoneTask.invokeTask (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:500:21)
    at Socket.ZoneTask.invoke (D:\3\azuredatastudio\node_modules\zone.js\dist\zone-node.js:485:48)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at Pipe.onread (net.js:594:20)

### PR DESCRIPTION
**Pull request by YurongHe**
Yesterday my account has issue. So I have to resend PR in order to get test execute.

The session is not ready yet when open and close notebook quickly. So clientSession used in NotebookModel.setProviderIdForKerneldispose is not set yet, which cause #4206:
TypeError: Cannot read property '0' of undefine.

And second issue in # 4207: ERR Cannot read property 'shutdownAll' of undefined: TypeError: Cannot read property 'shutdownAll' of undefined
at JupyterSessionManager.shutdownAll

### CODE REVIEW

File: src/sql/parts/notebook/models/clientSession.ts [hunk 1]
```diff
@@ -302,7 +302,7 @@ export class ClientSession implements IClientSession {
 	 */
 	public async shutdown(): Promise<void> {
 		// Always try to shut down session
-		if (this._session && this._session.id) {
+		if (this._session && this._session.id && this.notebookManager.sessionManager.isReady) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/models/clientSession.ts]: So I looked at the JupterSessionManager and it no-ops if session isn't ready anyhow. This one therefore shouldn't be doing anything / shouldn't be a cause of errors.

File: extensions/notebook/src/jupyter/jupyterNotebookManager.ts [hunk 2]
```diff
@@ -46,7 +46,7 @@ export class JupyterNotebookManager implements nb.NotebookManager, vscode.Dispos
 	}
 
 	dispose() {
-		if (this._sessionManager) {
+		if (this._sessionManager && this.sessionManager.isReady) {
```
Comment (kevcunnane) [inline extensions/notebook/src/jupyter/jupyterNotebookManager.ts]: For this one, can we fix in the `shutdownAll` method instead? That's what we do for regular shutdown where if not ready we auto-resolve the shutdown and don't call through to the APIs.

**Comment by kevcunnane** [comment]
One more thought on this - should we handle the case where in `setJupyterSessionManager` 's ready handler, it should check if `shutdownAll` was requested and if so should just immediately shut down?

File: src/sql/parts/notebook/models/clientSession.ts (hunk 1 above)
Comment (YurongHe) [reply src/sql/parts/notebook/models/clientSession.ts]: Thanks for checking this.

File: extensions/notebook/src/jupyter/jupyterNotebookManager.ts (hunk 2 above)
Comment (YurongHe) [reply extensions/notebook/src/jupyter/jupyterNotebookManager.ts]: @kevcunnane, I tried this. If removed this check and check isReday in JuptyerSessionManager, will hit The issue will happen to  this._sessionManager.dispose() in then.

[Microsoft.notebook]Cannot read property 'dispose' of undefined
console.(anonymous function) @ D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:2280
MainThreadExtensionService.$onExtensionRuntimeError @ mainThreadExtensionService.ts:42
RPCProtocol._doInvokeHandler @ rpcProtocol.ts:461
RPCProtocol._invokeHandler @ rpcProtocol.ts:446
RPCProtocol._receiveRequest @ rpcProtocol.ts:366
RPCProtocol._receiveOneMessage @ rpcProtocol.ts:296
(anonymous) @ rpcProtocol.ts:171
Emitter.fire @ event.ts:228
acceptChunk @ ipc.net.ts:103
Protocol._socketDataListener @ ipc.net.ts:132
ZoneDelegate.invokeTask @ D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:421
Zone.runTask @ D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:188
ZoneTask.invokeTask @ D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:496
ZoneTask.invoke @ D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:485
emitOne @ events.js:116
emit @ events.js:211
addChunk @ _stream_readable.js:263
readableAddChunk @ _stream_readable.js:250
Readable.push @ _stream_readable.js:208
onread @ net.js:594
D:\4\azuredatastudio\node_modules\zone.js\dist\zone-node.js:2280 TypeError: Cannot read property 'dispose' of undefined
	at JupyterSessionManager.dispose

**Pull request merged**